File: app/src/main/java/com/limhenry/androidthings/digitalclock/MainActivity.java
package com.limhenry.androidthings.digitalclock;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.hardware.display.DisplayManager;
import android.os.Bundle;
import android.os.Handler;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.Display;
import android.view.View;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.google.android.gms.vision.Frame;
import com.google.android.things.device.ScreenManager;

import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.TimeZone;

public class MainActivity extends Activity {
    boolean shouldExecuteOnResume;
    private static Context mContext;
    private OWMWeather owmWeather;
    private View.OnClickListener mWeatherListener = new View.OnClickListener() {
        public void onClick(View v) {
            owmWeather.getWeather();
        }
    };
    private Handler ambientModeHandler;
    private Runnable ambientModeRunnable;
    private CurrentPlaying currentPlaying;

    public static Context getContext() {
        return mContext;
    }

    public void startSpotify(View view) {
        ambientModeHandler.removeCallbacks(ambientModeRunnable);
        Intent intent = new Intent(this, SpotifyPlayerActivity.class);
        startActivity(intent);
    }

    public void startCalendar(View view) {
        ambientModeHandler.removeCallbacks(ambientModeRunnable);
        Intent intent = new Intent(this, CalendarActivity.class);
        startActivity(intent);
    }

    public void refreshMediaRoute(View view) {
        currentPlaying.refreshMediaRoute();
    }

    public void toggleAmbient(View view) {
        toggleAmbientMode();
    }

    public void toggleAmbientMode() {
        adjustBrightness(210);
        final LinearLayout layout_control = (LinearLayout) findViewById(R.id.layout_control);
        final LinearLayout layout_datetime = (LinearLayout) findViewById(R.id.layout_datetime);
        final ImageView ambient_bg = (ImageView) findViewById(R.id.ambient_bg);
        final LinearLayout.LayoutParams params = (LinearLayout.LayoutParams) layout_datetime.getLayoutParams();

        ambient_bg.setVisibility(View.INVISIBLE);
        layout_control.setVisibility(View.VISIBLE);
        params.setMargins(0, 64, 0, 0);
        layout_datetime.setLayoutParams(params);

        ambientModeHandler = new Handler();
        ambientModeRunnable = new Runnable() {
            public void run() {
                adjustBrightness(1);
                ambient_bg.setVisibility(View.VISIBLE);
                layout_control.setVisibility(View.GONE);
                params.setMargins(0, 0, 0, 0);
                layout_datetime.setLayoutParams(params);
            }
        };
        ambientModeHandler.postDelayed(ambientModeRunnable, 15000);
    }

    public void updateTime() {
        final SimpleDateFormat time = new SimpleDateFormat("hh:mm");
        final TextView txt_clockText = (TextView) findViewById(R.id.txt_clock_time);
        time.setTimeZone(TimeZone.getTimeZone("GMT+8"));
        txt_clockText.setText(time.format(new Date()));

        final SimpleDateFormat date = new SimpleDateFormat("EEEE, MMM dd");
        final TextView txt_clock_date = (TextView) findViewById(R.id.txt_clock_date);
        date.setTimeZone(TimeZone.getTimeZone("GMT+8"));
        txt_clock_date.setText(date.format(new Date()));

        final Handler handler = new Handler();
        handler.postDelayed(new Runnable() {
            @Override
            public void run() {
                txt_clockText.setText(time.format(new Date()));
                txt_clock_date.setText(date.format(new Date()));
                updateTime();
            }
        }, 30000);
    }

    public void adjustBrightness(int brightness) {
        DisplayManager displayManager = (DisplayManager) getApplicationContext().getSystemService(Context.DISPLAY_SERVICE);
        Display[] displays = displayManager.getDisplays();
        if (displays.length > 0) {
            int id = displays[0].getDisplayId();
            ScreenManager screenManager = new ScreenManager(id);
            screenManager.setBrightnessMode(ScreenManager.BRIGHTNESS_MODE_MANUAL);
            screenManager.setBrightness(brightness);
        }
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        mContext = getApplicationContext();

        owmWeather = new OWMWeather(this);
        owmWeather.getWeather();

        currentPlaying = new CurrentPlaying(this);

        updateTime();
        toggleAmbientMode();

        LinearLayout layout_time_weather = (LinearLayout) findViewById(R.id.layout_time_weather);
        layout_time_weather.setOnClickListener(mWeatherListener);

        LinearLayout main_linear = (LinearLayout) findViewById(R.id.main_linear);
        main_linear.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                ambientModeHandler.removeCallbacks(ambientModeRunnable);
                toggleAmbientMode();
            }
        });

        Handler handler = new Handler();
        handler.postDelayed(new Runnable() {
            public void run() {
                owmWeather.getWeather();
            }
        }, 300000);

        shouldExecuteOnResume = false;
    }

    @Override
    protected void onResume() {
        super.onResume();
        if(shouldExecuteOnResume){
            toggleAmbientMode();
        } else{
            shouldExecuteOnResume = true;
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
    }
}

File: app/src/main/java/com/limhenry/androidthings/digitalclock/MainActivity.kt
package com.limhenry.androidthings.digitalclock

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.graphics.Camera
import android.hardware.display.DisplayManager
import android.os.Bundle
import android.os.Handler
import android.provider.Settings
import android.util.Log
import android.view.Display
import android.view.View
import android.view.Window
import android.view.WindowManager
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView

import com.google.android.gms.tasks.OnCompleteListener
import com.google.android.gms.tasks.Task
import com.google.firebase.FirebaseApp
import com.google.firebase.auth.AuthResult
import com.google.firebase.auth.FirebaseAuth

import java.text.SimpleDateFormat
import java.util.Date
import java.util.TimeZone

class MainActivity : Activity() {
    internal var shouldExecuteOnResume = false
    private var owmWeather: OWMWeather? = null
    private var mWeatherListener: View.OnClickListener? = null
    private var ambientModeHandler: Handler? = null
    private var ambientModeRunnable: Runnable? = null
    private var mAuth: FirebaseAuth? = null
    private var homeAlarm: HomeAlarm? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        context = applicationContext

        //        currentPlaying = new CurrentPlaying(this);

        val main_linear = findViewById<LinearLayout>(R.id.main_linear)
        main_linear.setOnClickListener {
            ambientModeHandler!!.removeCallbacks(ambientModeRunnable)
            toggleAmbientMode()
        }

        val app_spotify = findViewById<LinearLayout>(R.id.app_spotify)
        app_spotify.setOnClickListener {
            ambientModeHandler!!.removeCallbacks(ambientModeRunnable)
            val intent = Intent(this@MainActivity, SpotifyPlayerActivity::class.java)
            startActivity(intent)
        }

        val app_selfie = findViewById<LinearLayout>(R.id.app_selfie)
        app_selfie.setOnClickListener {
//            val intent = Intent(this@MainActivity, CameraActivity::class.java)
//            startActivity(intent)
        }
    }

    fun refreshMediaRoute(view: View) {
        homeAlarm!!.setScanNetworkHandler()
    }

    fun toggleAmbientMode() {
//        adjustBrightness(210);
        val layout_control = findViewById<LinearLayout>(R.id.layout_control)
        val layout_datetime = findViewById<LinearLayout>(R.id.layout_datetime)
        val ambient_bg = findViewById<ImageView>(R.id.ambient_bg)
        val params = layout_datetime.layoutParams as LinearLayout.LayoutParams

        ambient_bg.visibility = View.INVISIBLE
        layout_control.visibility = View.VISIBLE
        params.setMargins(0, 64, 0, 0)
        layout_datetime.layoutParams = params

        ambientModeHandler = Handler()
        ambientModeRunnable = Runnable {
            //                adjustBrightness(1);
            ambient_bg.visibility = View.VISIBLE
            layout_control.visibility = View.GONE
            params.setMargins(0, 0, 0, 0)
            layout_datetime.layoutParams = params
        }
        ambientModeHandler!!.postDelayed(ambientModeRunnable, 15000)
    }

    fun updateTime() {
        val time = SimpleDateFormat("hh:mm")
        val txt_clockText = findViewById<TextView>(R.id.txt_clock_time)
        time.timeZone = TimeZone.getTimeZone("America/Chicago")
        txt_clockText.text = time.format(Date())

        val date = SimpleDateFormat("EEEE, MMM dd")
        val txt_clock_date = findViewById<TextView>(R.id.txt_clock_date)
        date.timeZone = TimeZone.getTimeZone("America/Chicago")
        txt_clock_date.text = date.format(Date())

        val handler = Handler()
        handler.postDelayed({
            txt_clockText.text = time.format(Date())
            txt_clock_date.text = date.format(Date())
            updateTime()
        }, 30000)
    }

    fun adjustBrightness(brightness: Int) {

        val displayManager = applicationContext.getSystemService(Context.DISPLAY_SERVICE) as DisplayManager
        val displays = displayManager.displays
        if (displays.size > 0) {
            // Change the screen brightness change mode to manual.
            Settings.System.putInt(contentResolver, Settings.System.SCREEN_BRIGHTNESS_MODE, Settings.System.SCREEN_BRIGHTNESS_MODE_MANUAL)
            // Apply the screen brightness value to the system, this will change the value in Settings ---> Display ---> Brightness level.
            // It will also change the screen brightness for the device.
            Settings.System.putInt(contentResolver, Settings.System.SCREEN_BRIGHTNESS, brightness)

            val window = window
            val layoutParams = window.attributes
            layoutParams.screenBrightness = brightness / 255f
            window.attributes = layoutParams


            //            int id = displays[0].getDisplayId();
            //            ScreenManager screenManager = new ScreenManager(id);
            //            screenManager.setBrightnessMode(ScreenManager.BRIGHTNESS_MODE_MANUAL);
            //            screenManager.setBrightness(brightness);
        }
    }

    override fun onResume() {
        super.onResume()
        if (shouldExecuteOnResume) {
            toggleAmbientMode()
        } else {
            owmWeather = OWMWeather(this)
            owmWeather!!.getWeather()

            FirebaseApp.initializeApp(this)
            mAuth = FirebaseAuth.getInstance()

            mAuth!!.signInWithEmailAndPassword(getString(R.string.firebase_email_test), getString(R.string.firebase_password_test))
                    .addOnCompleteListener(this) { task ->
                        if (task.isSuccessful) {
                            homeAlarm = HomeAlarm(this@MainActivity)
                        }
                    }

            mWeatherListener = View.OnClickListener { owmWeather!!.getWeather() }

            updateTime()
            toggleAmbientMode()

            val layout_time_weather = findViewById<LinearLayout>(R.id.layout_time_weather)
            layout_time_weather.setOnClickListener(mWeatherListener)

            val handler = Handler()
            handler.postDelayed({ owmWeather!!.getWeather() }, 150000)

            shouldExecuteOnResume = true
        }
    }

    override fun onDestroy() {
        super.onDestroy()
    }

    companion object {
        public var context: Context? = null
            private set
    }
}